# Manipulating and Analyzing Large Data using Pandas

# DataSet 

## https://www.kaggle.com/datasets/josephcheng123456/olympic-historical-dataset-from-olympediaorg

## 154,902 unique athletes, country they belong to, medals they won and their biological information i.e. height, weight, date of birth
## All Winter / Summer Olympic games from 1896 to 2022

# Importing, Indexing, Editing, Combining, Grouping, and Aggregating DataFrames.

## Import DataFrames

In [42]:
import pandas as pd
df = pd.read_csv("Olympic_Athlete_Bio.csv")
df.head()

# .head() is a method
# .read_csv() is a method in pandas, specifically a method of the pandas library's DataFrame class.


athlete_id                 name     sex        born height weight  \
0       16809       Károly Teppert    Male  1891-07-20     na     na   
1       43737    Andrzej Socharski    Male  1947-08-31  173.0     72   
2       50147  Nathalie Wunderlich  Female  1971-06-03  170.0     50   
3        5085           Miha Lokar    Male  1935-09-10  182.0     76   
4      136329          Austin Hack    Male  1992-05-17  203.0    100   

          country country_noc  \
0         Hungary         HUN   
1          Poland         POL   
2     Switzerland         SUI   
3      Yugoslavia         YUG   
4   United States         USA   

                                         description  \
0  Károly Teppert started competing in cycling in...   
1                                                 na   
2                                                 na   
3                                                 na   
4                                                 na   

                                       special_notes  
0                                                 na  
1  Listed in Olympians Who Won a Medal at the Wor...  
2                                                 na  
3  Listed in Olympians Who Won a Medal at the Eur...  
4                                                 na

In [43]:
print("Number of rows: {}, number of columns: {}".format(df.shape[0],df.shape[1]))
print(f"Number of rows: {df.shape[0]}, number of columns: {df.shape[1]}")

# The format() function is used to print the number of rows and columns in a DataFrame df. 
# The df.shape attribute returns a tuple representing the dimensions of the DataFrame, 
# where the first element is the number of rows and the second element is the number of columns. 
# The format() function is then used to insert these values into the string template.

# Here's what the code does step by step:

# df.shape[0] retrieves the number of rows in the DataFrame df.
# df.shape[1] retrieves the number of columns in the DataFrame df.
# The format() function replaces the placeholders {} in the string with the respective values 
# obtained from the previous steps.
# The formatted string is then printed to the console.

Number of rows: 155031, number of columns: 10
Number of rows: 155031, number of columns: 10


## Store the column with the name of the athletes as one Data Frame

In [44]:
names = df["name"]
print(names)


0                     Károly Teppert
1                  Andrzej Socharski
2                Nathalie Wunderlich
3                         Miha Lokar
4                        Austin Hack
                     ...            
155026                  Jean Palluch
155027    Hendrik Mangelaar Meertens
155028                 Jo Hyeong-Won
155029                   Raúl Maroto
155030                Shizo Kanakuri
Name: name, Length: 155031, dtype: object


# Print the top 5 names

In [45]:
names.head()

0         Károly Teppert
1      Andrzej Socharski
2    Nathalie Wunderlich
3             Miha Lokar
4            Austin Hack
Name: name, dtype: object

# Store the columns with the names of the athletes as one Data Frame along with height and weight 

In [46]:
dimensions = df[["name","height", "weight"]]
dimensions.head()

name height weight
0       Károly Teppert     na     na
1    Andrzej Socharski  173.0     72
2  Nathalie Wunderlich  170.0     50
3           Miha Lokar  182.0     76
4          Austin Hack  203.0    100

# Read the second dataset 

In [47]:
df1 = pd.read_csv("Olympic_Athlete_Event_Results.csv")
df1.head()

edition  edition_id country_noc      sport            event  \
0  1908 Summer Olympics           5         ANZ  Athletics  100 metres, Men   
1  1908 Summer Olympics           5         ANZ  Athletics  400 metres, Men   
2  1908 Summer Olympics           5         ANZ  Athletics  800 metres, Men   
3  1908 Summer Olympics           5         ANZ  Athletics  800 metres, Men   
4  1908 Summer Olympics           5         ANZ  Athletics  800 metres, Men   

   result_id          athlete  athlete_id        pos medal  isTeamSport  
0      56265  Ernest Hutcheon       64710        DNS    na        False  
1      56313     Henry Murray       64756        DNS    na        False  
2      56338    Harvey Sutton       64808  3 h8 r1/2    na        False  
3      56338      Guy Haskins      922519        DNS    na        False  
4      56338     Joseph Lynch       64735        DNS    na        False

# Print the number of rows and colums in the second dataset

In [48]:
print("Number of rows: {}, number of columns: {}".format(df.shape[0],df.shape[1]))

Number of rows: 155031, number of columns: 10


# Print the last 5 rows

In [49]:
df1.tail()

edition  edition_id country_noc      sport  \
314902  2022 Winter Olympics          62         NED  Bobsleigh   
314903  2022 Winter Olympics          62         SVK  Bobsleigh   
314904  2022 Winter Olympics          62         KOR  Bobsleigh   
314905  2022 Winter Olympics          62         JAM  Bobsleigh   
314906  2022 Winter Olympics          62         UKR  Bobsleigh   

                 event  result_id                     athlete  athlete_id pos  \
314902  Monobob, Women   19019671              Karlien Sleper      148662  16   
314903  Monobob, Women   19019671          Viktória Čerňanská      138683  17   
314904  Monobob, Women   19019671                  Kim Yu-Ran      137542  18   
314905  Monobob, Women   19019671  Jazmine Fenlator-Victorian      128682  19   
314906  Monobob, Women   19019671                Lidiya Hunko      149047  20   

       medal  isTeamSport  
314902    na        False  
314903    na        False  
314904    na        False  
314905    na        False  
314906    na        False

# Print the last 10 rows

In [50]:
df1.tail(10)

edition  edition_id country_noc      sport  \
314897  2022 Winter Olympics          62         FRA  Bobsleigh   
314898  2022 Winter Olympics          62         ROU  Bobsleigh   
314899  2022 Winter Olympics          62         GER  Bobsleigh   
314900  2022 Winter Olympics          62         AUT  Bobsleigh   
314901  2022 Winter Olympics          62         ITA  Bobsleigh   
314902  2022 Winter Olympics          62         NED  Bobsleigh   
314903  2022 Winter Olympics          62         SVK  Bobsleigh   
314904  2022 Winter Olympics          62         KOR  Bobsleigh   
314905  2022 Winter Olympics          62         JAM  Bobsleigh   
314906  2022 Winter Olympics          62         UKR  Bobsleigh   

                 event  result_id                     athlete  athlete_id pos  \
314897  Monobob, Women   19019671                 Margot Boch     2301119  11   
314898  Monobob, Women   19019671               Andreea Grecu      128289  12   
314899  Monobob, Women   19019671             Mariama Jamanka      137253  13   
314900  Monobob, Women   19019671                 Kati Beierl      136713  14   
314901  Monobob, Women   19019671             Giada Andreutti      148470  15   
314902  Monobob, Women   19019671              Karlien Sleper      148662  16   
314903  Monobob, Women   19019671          Viktória Čerňanská      138683  17   
314904  Monobob, Women   19019671                  Kim Yu-Ran      137542  18   
314905  Monobob, Women   19019671  Jazmine Fenlator-Victorian      128682  19   
314906  Monobob, Women   19019671                Lidiya Hunko      149047  20   

       medal  isTeamSport  
314897    na        False  
314898    na        False  
314899    na        False  
314900    na        False  
314901    na        False  
314902    na        False  
314903    na        False  
314904    na        False  
314905    na        False  
314906    na        False

## Need a dataframe which will provide the Gold winners , edition and country_noc

In [51]:
# Filter for Gold medal winners
gold_winners_df = df1[df1['medal'] == 'Gold']

# Select the desired columns
desired_columns = ['edition', 'edition_id', 'country_noc', 'sport', 'event', 'result_id', 'athlete', 'athlete_id', 'pos', 'medal', 'isTeamSport']
gold_winners_df = gold_winners_df[desired_columns]

# Reset the index of the new DataFrame
gold_winners_df.reset_index(drop=True, inplace=True)

## gold_winners_df.reset_index(drop=True, inplace=True) is used to reset the index of 
## gold_winners_df, removing the old index column and applying the changes directly to the same DataFrame object.

# Display the new DataFrame
gold_winners_df.head()

edition  edition_id country_noc  sport       event  result_id  \
0  1908 Summer Olympics           5         ANZ  Rugby  Rugby, Men      31505   
1  1908 Summer Olympics           5         ANZ  Rugby  Rugby, Men      31505   
2  1908 Summer Olympics           5         ANZ  Rugby  Rugby, Men      31505   
3  1908 Summer Olympics           5         ANZ  Rugby  Rugby, Men      31505   
4  1908 Summer Olympics           5         ANZ  Rugby  Rugby, Men      31505   

           athlete  athlete_id pos medal  isTeamSport  
0     John Barnett       11237   1  Gold         True  
1  Phil Carmichael       11239   1  Gold         True  
2      Dan Carroll       11240   1  Gold         True  
3        Bob Craig       11241   1  Gold         True  
4      Tom Griffin       11242   1  Gold         True

# Replace Gold with G 

In [52]:
gold_winners_df["medal_short"] = gold_winners_df.medal.map(lambda x: "G" if x == "Gold" else "None") 

# medal_short is the column that we want to introduce
# .map(lambda x: "G" if x == "Gold" else "None"): This is a transformation applied to each value in the 
# "medal" column using the map() function and a lambda (anonymous) function.

# lambda x: "G" if x == "Gold" else "None": This is a lambda function that takes a single argument x, 
# which represents each value in the "medal" column. If the value is equal to "Gold", it returns "G"; otherwise, 
# it returns "None". This is a way to create a new column with a simplified representation of the medals, where "Gold" 
# is represented as "G" and all other values are represented as "None".
# gold_winners_df["medal_short"] = ...: This assigns the result of the mapping operation (either "G" or "None") 
# to the new "medal_short" column in the gold_winners_df DataFrame.

gold_winners_df.head()

edition  edition_id country_noc  sport       event  result_id  \
0  1908 Summer Olympics           5         ANZ  Rugby  Rugby, Men      31505   
1  1908 Summer Olympics           5         ANZ  Rugby  Rugby, Men      31505   
2  1908 Summer Olympics           5         ANZ  Rugby  Rugby, Men      31505   
3  1908 Summer Olympics           5         ANZ  Rugby  Rugby, Men      31505   
4  1908 Summer Olympics           5         ANZ  Rugby  Rugby, Men      31505   

           athlete  athlete_id pos medal  isTeamSport medal_short  
0     John Barnett       11237   1  Gold         True           G  
1  Phil Carmichael       11239   1  Gold         True           G  
2      Dan Carroll       11240   1  Gold         True           G  
3        Bob Craig       11241   1  Gold         True           G  
4      Tom Griffin       11242   1  Gold         True           G

# New Data Frame 10th to 210th row, Sport to Athlete 

## Use of .loc attribute

### An attribute typically refers to a value or property associated with an object or data structure. It provides information about the object's characteristics, state, or behavior. Attributes can be thought of as variables that are associated with objects, and they can have different types and values.

## .loc[10:210, "sport":"athlete"]: This is the indexing operation used to select a 
## subset of rows and columns from the DataFrame.
## .loc is a method in pandas

In [53]:
gold_winners_df1 = gold_winners_df.loc[10:210, "sport":"athlete"]
gold_winners_df1.head()

sport       event  result_id            athlete
10  Rugby  Rugby, Men      31505  Charles McMurtrie
11  Rugby  Rugby, Men      31505      Syd Middleton
12  Rugby  Rugby, Men      31505       Tom Richards
13  Rugby  Rugby, Men      31505    Charles Russell
14  Rugby  Rugby, Men      31505        Frank Smith

In [54]:
gold_winners_df1.reset_index(drop=True, inplace=True)

#.reset_index(): This function is used to reset the index of a DataFrame. 
#The index is a way to label and identify rows in a DataFrame.

# drop=True: This is an argument passed to the reset_index() function. 
# When drop=True, it means that the current index will be removed from the DataFrame 
# and a default integer index will be assigned instead. 

#inplace=True: This is another argument passed to the reset_index() function. 
# When inplace=True, it means that the changes will be applied directly to the DataFrame gold_winners_df1 
# without creating a new DataFrame. 
# If inplace=False (or omitted), 
# the function would return a new DataFrame with the index reset, but the original DataFrame would remain unchanged.

gold_winners_df1.head()

sport       event  result_id            athlete
0  Rugby  Rugby, Men      31505  Charles McMurtrie
1  Rugby  Rugby, Men      31505      Syd Middleton
2  Rugby  Rugby, Men      31505       Tom Richards
3  Rugby  Rugby, Men      31505    Charles Russell
4  Rugby  Rugby, Men      31505        Frank Smith

# Create a dataframe with index 0, 20, 40 in rows and 2, 6, 9 columns

# Use of iloc method 


## iloc is not a method or function in the traditional sense, but rather it is an attribute used to perform integer-based indexing on pandas DataFrames and Series. It stands for "integer location" and is used to access rows and columns of a DataFrame using integer indices rather than labels.

# .loc vs .iloc

### Both .loc and .iloc are attributes in pandas, not methods.

#### An attribute typically refers to a value or property associated with an object or data structure. It provides information about the object's characteristics, state, or behavior. Attributes can be thought of as variables that are associated with objects, and they can have different types and values.

### .loc: This is used for label-based indexing. It allows you to access a group of rows and columns 
###    by labels or a boolean array. It's used in the form dataframe.loc[row_label, column_label]. 
###    It's not a method; rather, it's a way to access specific data based on labels.

### .iloc: This is used for integer-location based indexing. It allows you to access a group of rows and columns
###    by integer position. It's used in the form dataframe.iloc[row_index, column_index]. 
###    It's also not a method; it's a way to access data based on its integer position.

In [55]:
gold_winners_df2 = gold_winners_df.iloc[[10, 20, 40], [2,6,9]]
gold_winners_df2.head()

country_noc                   athlete medal
10         ANZ         Charles McMurtrie  Gold
20         ANZ             Ward Prentice  Gold
40         AZE  Zemfira Meftəkhətddinova  Gold

# Combine df and df1. Find Olympians with weight > 100 

In [56]:
df.head()

athlete_id                 name     sex        born height weight  \
0       16809       Károly Teppert    Male  1891-07-20     na     na   
1       43737    Andrzej Socharski    Male  1947-08-31  173.0     72   
2       50147  Nathalie Wunderlich  Female  1971-06-03  170.0     50   
3        5085           Miha Lokar    Male  1935-09-10  182.0     76   
4      136329          Austin Hack    Male  1992-05-17  203.0    100   

          country country_noc  \
0         Hungary         HUN   
1          Poland         POL   
2     Switzerland         SUI   
3      Yugoslavia         YUG   
4   United States         USA   

                                         description  \
0  Károly Teppert started competing in cycling in...   
1                                                 na   
2                                                 na   
3                                                 na   
4                                                 na   

                                       special_notes  
0                                                 na  
1  Listed in Olympians Who Won a Medal at the Wor...  
2                                                 na  
3  Listed in Olympians Who Won a Medal at the Eur...  
4                                                 na

In [57]:
df1.head()

edition  edition_id country_noc      sport            event  \
0  1908 Summer Olympics           5         ANZ  Athletics  100 metres, Men   
1  1908 Summer Olympics           5         ANZ  Athletics  400 metres, Men   
2  1908 Summer Olympics           5         ANZ  Athletics  800 metres, Men   
3  1908 Summer Olympics           5         ANZ  Athletics  800 metres, Men   
4  1908 Summer Olympics           5         ANZ  Athletics  800 metres, Men   

   result_id          athlete  athlete_id        pos medal  isTeamSport  
0      56265  Ernest Hutcheon       64710        DNS    na        False  
1      56313     Henry Murray       64756        DNS    na        False  
2      56338    Harvey Sutton       64808  3 h8 r1/2    na        False  
3      56338      Guy Haskins      922519        DNS    na        False  
4      56338     Joseph Lynch       64735        DNS    na        False

In [58]:
import pandas as pd

# Assuming df1 and df2 are your existing DataFrames

# Step 1: Concatenate df1 and df2
combined_df = pd.concat([df, df1], ignore_index=True)

combined_df.head()






athlete_id                 name     sex        born height weight  \
0       16809       Károly Teppert    Male  1891-07-20     na     na   
1       43737    Andrzej Socharski    Male  1947-08-31  173.0     72   
2       50147  Nathalie Wunderlich  Female  1971-06-03  170.0     50   
3        5085           Miha Lokar    Male  1935-09-10  182.0     76   
4      136329          Austin Hack    Male  1992-05-17  203.0    100   

          country country_noc  \
0         Hungary         HUN   
1          Poland         POL   
2     Switzerland         SUI   
3      Yugoslavia         YUG   
4   United States         USA   

                                         description  \
0  Károly Teppert started competing in cycling in...   
1                                                 na   
2                                                 na   
3                                                 na   
4                                                 na   

                                       special_notes edition  edition_id  \
0                                                 na     NaN         NaN   
1  Listed in Olympians Who Won a Medal at the Wor...     NaN         NaN   
2                                                 na     NaN         NaN   
3  Listed in Olympians Who Won a Medal at the Eur...     NaN         NaN   
4                                                 na     NaN         NaN   

  sport event  result_id athlete  pos medal isTeamSport  
0   NaN   NaN        NaN     NaN  NaN   NaN         NaN  
1   NaN   NaN        NaN     NaN  NaN   NaN         NaN  
2   NaN   NaN        NaN     NaN  NaN   NaN         NaN  
3   NaN   NaN        NaN     NaN  NaN   NaN         NaN  
4   NaN   NaN        NaN     NaN  NaN   NaN         NaN

In [85]:
# Convert 'weight' column to numeric
combined_df['weight'] = pd.to_numeric(combined_df['weight'], errors='coerce')

# Remove rows with weight = NaN
weight = combined_df.dropna(subset=['weight'])

# Display the cleaned and filtered DataFrame
weight.tail()

athlete_id            name     sex        born height  weight  \
155025       74088  Oddrun Hokland  Female  1942-11-29  162.0    59.0   
155026       25136    Jean Palluch    Male  1923-12-23  180.0    80.0   
155028       11757   Jo Hyeong-Won    Male  1936-07-12  173.0    65.0   
155029       21666     Raúl Maroto    Male  1965-09-05  190.0    76.0   
155030       72486  Shizo Kanakuri    Male  1891-08-20  170.0    64.0   

                   country country_noc  \
155025              Norway         NOR   
155026              France         FRA   
155028   Republic of Korea         KOR   
155029               Spain         ESP   
155030               Japan         JPN   

                                              description  \
155025  Personal Bests: LJ – 6.26 (1964); Pen – 4540 (...   
155026                                                 na   
155028                                                 na   
155029                                                 na   
155030  Kanakuri had a weak constitution as a child un...   

                                            special_notes edition  edition_id  \
155025                                                 na     NaN         NaN   
155026                                                 na     NaN         NaN   
155028                                                 na     NaN         NaN   
155029  Listed in Olympians Who Won a Medal at the Med...     NaN         NaN   
155030                                                 na     NaN         NaN   

       sport event  result_id athlete  pos medal isTeamSport  
155025   NaN   NaN        NaN     NaN  NaN   NaN         NaN  
155026   NaN   NaN        NaN     NaN  NaN   NaN         NaN  
155028   NaN   NaN        NaN     NaN  NaN   NaN         NaN  
155029   NaN   NaN        NaN     NaN  NaN   NaN         NaN  
155030   NaN   NaN        NaN     NaN  NaN   NaN         NaN

In [84]:
# Filter for Gold winners with weight > 100
Healthy_olympians = weight[weight['weight'] > 100]

# Display the filtered DataFrame
Healthy_olympians.tail()

athlete_id                   name   sex        born height  weight  \
154898       58584           György Kékes  Male  1966-06-21  191.0   125.0   
154948       41094  Manuel José Fernández  Male  1927-06-10  181.0   105.0   
154961      121142         Cristi Pîrghie  Male  1992-07-20  202.0   108.0   
154987       76585        Jaroslav Brabec  Male  1949-07-27  190.0   125.0   
155021       56961           Mark Cameron  Male  1952-10-10  178.0   109.0   

                country country_noc  \
154898          Hungary         HUN   
154948        Argentina         ARG   
154961          Romania         ROU   
154987   Czechoslovakia         TCH   
155021    United States         USA   

                                              description  \
154898  György Kékes had the following finishes at maj...   
154948                                                 na   
154961                                                 na   
154987  Jaroslav Brabec was a shot putter who competed...   
155021  Mark Cameron was a six-time AAU Champion from ...   

                                            special_notes edition  edition_id  \
154898                                                 na     NaN         NaN   
154948  Listed in Olympians Who Won a Medal at the Sum...     NaN         NaN   
154961                                                 na     NaN         NaN   
154987  Listed in Olympians Who Won a Medal at the Eur...     NaN         NaN   
155021  Listed in Olympians Who Won a Medal at the Sum...     NaN         NaN   

       sport event  result_id athlete  pos medal isTeamSport  
154898   NaN   NaN        NaN     NaN  NaN   NaN         NaN  
154948   NaN   NaN        NaN     NaN  NaN   NaN         NaN  
154961   NaN   NaN        NaN     NaN  NaN   NaN         NaN  
154987   NaN   NaN        NaN     NaN  NaN   NaN         NaN  
155021   NaN   NaN        NaN     NaN  NaN   NaN         NaN

# Inner Join, Left Join, Right Join and Outer Join 

## These are terms commonly used in relational databases and SQL to describe different types of data combining operations. These operations involve merging data from two or more tables based on a common column (usually known as a key).


## Inner Join:
## An inner join returns only the rows where there is a match in both tables based on the specified key. It keeps only the common elements between both tables.

## Left join
## A left join returns all the rows from the left table and the matching rows from the right table based on the specified key. If there's no match in the right table, the result will still include the row from the left table, but with NULL values for the columns from the right table.

## Right Join
## A right join is similar to a left join, but it returns all the rows from the right table and the matching rows from the left table. If there's no match in the left table, the result will still include the row from the right table, but with NULL values for the columns from the left table.

## Outer Join
## An outer join is a type of join that includes all the rows from both tables, along with the matching rows based on the specified key. 


In [62]:
df1.head()

edition  edition_id country_noc      sport            event  \
0  1908 Summer Olympics           5         ANZ  Athletics  100 metres, Men   
1  1908 Summer Olympics           5         ANZ  Athletics  400 metres, Men   
2  1908 Summer Olympics           5         ANZ  Athletics  800 metres, Men   
3  1908 Summer Olympics           5         ANZ  Athletics  800 metres, Men   
4  1908 Summer Olympics           5         ANZ  Athletics  800 metres, Men   

   result_id          athlete  athlete_id        pos medal  isTeamSport  
0      56265  Ernest Hutcheon       64710        DNS    na        False  
1      56313     Henry Murray       64756        DNS    na        False  
2      56338    Harvey Sutton       64808  3 h8 r1/2    na        False  
3      56338      Guy Haskins      922519        DNS    na        False  
4      56338     Joseph Lynch       64735        DNS    na        False

## Let's split df1 into two Data Frames keeping athlete column common in both Data Frames 

In [86]:
df1_split1 = df1.iloc[0:155031, 0:7]
print("Number of rows: {}, number of columns: {}".format(df1_split1.shape[0],df1_split1.shape[1]))
df1_split1.tail()

Number of rows: 155031, number of columns: 7


edition  edition_id country_noc             sport  \
155026  2018 Winter Olympics          60         JPN  Freestyle Skiing   
155027  2018 Winter Olympics          60         JPN  Freestyle Skiing   
155028  2018 Winter Olympics          60         JPN  Freestyle Skiing   
155029  2018 Winter Olympics          60         JPN        Ice Hockey   
155030  2018 Winter Olympics          60         JPN        Ice Hockey   

                    event  result_id        athlete  
155026    Halfpipe, Women    9000568  Ayana Onozuka  
155027    Halfpipe, Women    9000568   Saori Suzuki  
155028    Halfpipe, Women    9000568   Yurie Watabe  
155029  Ice Hockey, Women    9000700  Nana Fujimoto  
155030  Ice Hockey, Women    9000700   Shiori Koike

In [87]:
df1_split2 = df1.iloc[0:155031, 6:11]
print("Number of rows: {}, number of columns: {}".format(df1_split2.shape[0],df1_split2.shape[1]))
df1_split2.tail()

Number of rows: 155031, number of columns: 5


athlete  athlete_id pos medal  isTeamSport
155026  Ayana Onozuka      127983   5   NaN        False
155027   Saori Suzuki      137448  14   NaN        False
155028   Yurie Watabe      137450  22   NaN        False
155029  Nana Fujimoto      127994   6   NaN         True
155030   Shiori Koike      127997   6   NaN         True

# For df1_split2 we replace the athlete names as NAN from index 0 to 155029

In [65]:
import numpy as np
df1_split2.loc[0:155020, 'athlete'] = np.nan
df1_split2.tail(30)

athlete  athlete_id      pos   medal  isTeamSport
155001              NaN      137460        2  Silver        False
155002              NaN      137459       18      na        False
155003              NaN      137455        4      na        False
155004              NaN      137457        6      na        False
155005              NaN      137458  21 r1/2      na         True
155006              NaN      127989  21 r1/2      na         True
155007              NaN      137456       15      na         True
155008              NaN      119065       15      na         True
155009              NaN      127988        5      na         True
155010              NaN      137459        5      na         True
155011              NaN      137460        5      na         True
155012              NaN      137455        5      na         True
155013              NaN      137457        5      na         True
155014              NaN      137458        5      na         True
155015              NaN      127989        5      na         True
155016              NaN      137456        5      na         True
155017              NaN      119065        5      na         True
155018              NaN      137451        3  Bronze        False
155019              NaN      119025       10      na        False
155020              NaN      137452       11      na        False
155021   Nobuyuki Nishi      119005       19      na        False
155022     Naoya Tabara      137453       19      na        False
155023  Taisei Yamamoto      137454       20      na        False
155024     Arisa Murata      118944       18      na        False
155025    Reina Umehara      137449       22      na        False
155026    Ayana Onozuka      127983        5      na        False
155027     Saori Suzuki      137448       14      na        False
155028     Yurie Watabe      137450       22      na        False
155029    Nana Fujimoto      127994        6      na         True
155030     Shiori Koike      127997        6      na         True

# Inner Join

### An inner join returns only the rows where there is a match in both tables based on the specified key. It keeps only the common elements between both tables.


In [66]:
merged_df_inner = df1_split1.merge(df1_split2, on='athlete', how='inner')

merged_df_inner.tail()

edition  edition_id country_noc             sport  \
12  2018 Winter Olympics          60         JPN  Freestyle Skiing   
13  2018 Winter Olympics          60         JPN  Freestyle Skiing   
14  2018 Winter Olympics          60         JPN  Freestyle Skiing   
15  2018 Winter Olympics          60         JPN  Freestyle Skiing   
16  2018 Winter Olympics          60         JPN  Freestyle Skiing   

               event  result_id          athlete  athlete_id pos medal  \
12      Aerials, Men    9000508     Naoya Tabara      137453  19    na   
13   Slopestyle, Men    9000523  Taisei Yamamoto      137454  20    na   
14  Ski Cross, Women    9000582    Reina Umehara      137449  22    na   
15   Halfpipe, Women    9000568     Saori Suzuki      137448  14    na   
16   Halfpipe, Women    9000568     Yurie Watabe      137450  22    na   

    isTeamSport  
12        False  
13        False  
14        False  
15        False  
16        False

In [67]:
print("Number of rows: {}, number of columns: {}".format(merged_df_inner.shape[0],merged_df_inner.shape[1]))

Number of rows: 17, number of columns: 11


# Left Join

### A left join returns all the rows from the left table and the matching rows from the right table based on the specified key. If there's no match in the right table, the result will still include the row from the left table, but with NULL values for the columns from the right table.

In [68]:
merged_df_left = df1_split1.merge(df1_split2, on='athlete', how='left')

merged_df_left.tail()

edition  edition_id country_noc             sport  \
155026  2018 Winter Olympics          60         JPN  Freestyle Skiing   
155027  2018 Winter Olympics          60         JPN  Freestyle Skiing   
155028  2018 Winter Olympics          60         JPN  Freestyle Skiing   
155029  2018 Winter Olympics          60         JPN        Ice Hockey   
155030  2018 Winter Olympics          60         JPN        Ice Hockey   

                    event  result_id        athlete  athlete_id pos medal  \
155026    Halfpipe, Women    9000568  Ayana Onozuka    127983.0   5    na   
155027    Halfpipe, Women    9000568   Saori Suzuki    137448.0  14    na   
155028    Halfpipe, Women    9000568   Yurie Watabe    137450.0  22    na   
155029  Ice Hockey, Women    9000700  Nana Fujimoto    127994.0   6    na   
155030  Ice Hockey, Women    9000700   Shiori Koike    127997.0   6    na   

       isTeamSport  
155026       False  
155027       False  
155028       False  
155029        True  
155030        True

In [69]:
print("Number of rows: {}, number of columns: {}".format(merged_df_left.shape[0],merged_df_left.shape[1]))

Number of rows: 155031, number of columns: 11


# Right Join 

### A right join is similar to a left join, but it returns all the rows from the right table and the matching rows from the left table. If there's no match in the left table, the result will still include the row from the right table, but with NULL values for the columns from the left table.


In [70]:
merged_df_right = df1_split1.merge(df1_split2, on='athlete', how='right')

merged_df_right.tail()

edition  edition_id country_noc             sport  \
155033  2018 Winter Olympics        60.0         JPN  Freestyle Skiing   
155034  2014 Winter Olympics        58.0         JPN        Ice Hockey   
155035  2018 Winter Olympics        60.0         JPN        Ice Hockey   
155036  2014 Winter Olympics        58.0         JPN        Ice Hockey   
155037  2018 Winter Olympics        60.0         JPN        Ice Hockey   

                    event  result_id        athlete  athlete_id pos medal  \
155033    Halfpipe, Women  9000568.0   Yurie Watabe      137450  22    na   
155034  Ice Hockey, Women   350374.0  Nana Fujimoto      127994   6    na   
155035  Ice Hockey, Women  9000700.0  Nana Fujimoto      127994   6    na   
155036  Ice Hockey, Women   350374.0   Shiori Koike      127997   6    na   
155037  Ice Hockey, Women  9000700.0   Shiori Koike      127997   6    na   

        isTeamSport  
155033        False  
155034         True  
155035         True  
155036         True  
155037         True

In [71]:
print("Number of rows: {}, number of columns: {}".format(merged_df_right.shape[0],merged_df_right.shape[1]))

Number of rows: 155038, number of columns: 11


# Outer Join

## An outer join is a type of join that includes all the rows from both tables, along with the matching rows based on the specified key.

In [72]:
merged_df_outer = df1_split1.merge(df1_split2, on='athlete', how='outer')

merged_df_outer.tail(30)

edition  edition_id country_noc sport event  result_id athlete  \
310022     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310023     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310024     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310025     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310026     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310027     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310028     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310029     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310030     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310031     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310032     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310033     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310034     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310035     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310036     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310037     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310038     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310039     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310040     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310041     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310042     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310043     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310044     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310045     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310046     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310047     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310048     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310049     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310050     NaN         NaN         NaN   NaN   NaN        NaN     NaN   
310051     NaN         NaN         NaN   NaN   NaN        NaN     NaN   

        athlete_id      pos   medal isTeamSport  
310022    137446.0        8      na        True  
310023    137447.0        8      na        True  
310024    137444.0        8      na        True  
310025   1800049.0        8      na        True  
310026    137440.0        3  Bronze        True  
310027    127980.0        3  Bronze        True  
310028    137441.0        3  Bronze        True  
310029    137442.0        3  Bronze        True  
310030    110292.0        3  Bronze        True  
310031    127988.0        1    Gold       False  
310032    137460.0        2  Silver       False  
310033    137459.0       18      na       False  
310034    137455.0        4      na       False  
310035    137457.0        6      na       False  
310036    137458.0  21 r1/2      na        True  
310037    127989.0  21 r1/2      na        True  
310038    137456.0       15      na        True  
310039    119065.0       15      na        True  
310040    127988.0        5      na        True  
310041    137459.0        5      na        True  
310042    137460.0        5      na        True  
310043    137455.0        5      na        True  
310044    137457.0        5      na        True  
310045    137458.0        5      na        True  
310046    127989.0        5      na        True  
310047    137456.0        5      na        True  
310048    119065.0        5      na        True  
310049    137451.0        3  Bronze       False  
310050    119025.0       10      na       False  
310051    137452.0       11      na       False

In [73]:
print("Number of rows: {}, number of columns: {}".format(merged_df_outer.shape[0],merged_df_outer.shape[1]))

Number of rows: 310052, number of columns: 11


In [74]:
df.head()

athlete_id                 name     sex        born height weight  \
0       16809       Károly Teppert    Male  1891-07-20     na     na   
1       43737    Andrzej Socharski    Male  1947-08-31  173.0     72   
2       50147  Nathalie Wunderlich  Female  1971-06-03  170.0     50   
3        5085           Miha Lokar    Male  1935-09-10  182.0     76   
4      136329          Austin Hack    Male  1992-05-17  203.0    100   

          country country_noc  \
0         Hungary         HUN   
1          Poland         POL   
2     Switzerland         SUI   
3      Yugoslavia         YUG   
4   United States         USA   

                                         description  \
0  Károly Teppert started competing in cycling in...   
1                                                 na   
2                                                 na   
3                                                 na   
4                                                 na   

                                       special_notes  
0                                                 na  
1  Listed in Olympians Who Won a Medal at the Wor...  
2                                                 na  
3  Listed in Olympians Who Won a Medal at the Eur...  
4                                                 na

In [75]:
df1.head()

edition  edition_id country_noc      sport            event  \
0  1908 Summer Olympics           5         ANZ  Athletics  100 metres, Men   
1  1908 Summer Olympics           5         ANZ  Athletics  400 metres, Men   
2  1908 Summer Olympics           5         ANZ  Athletics  800 metres, Men   
3  1908 Summer Olympics           5         ANZ  Athletics  800 metres, Men   
4  1908 Summer Olympics           5         ANZ  Athletics  800 metres, Men   

   result_id          athlete  athlete_id        pos medal  isTeamSport  
0      56265  Ernest Hutcheon       64710        DNS    na        False  
1      56313     Henry Murray       64756        DNS    na        False  
2      56338    Harvey Sutton       64808  3 h8 r1/2    na        False  
3      56338      Guy Haskins      922519        DNS    na        False  
4      56338     Joseph Lynch       64735        DNS    na        False

# RESHAPING DATA

# Country as the index, columns athletes' names and values edition_id for those who have won gold

# Reshape data using .pivot()

In [76]:
# Replace 'na' with NaN
df1.replace('na', np.nan, inplace=True)

# Drop rows with any NaN values
df1_cleaned = df1.dropna()


# Convert 'edition_id' column to numeric, replacing non-convertible values with NaN
df1_cleaned['edition_id'] = pd.to_numeric(df1_cleaned['edition_id'], errors='coerce')

# Drop rows where 'edition_id' is NaN (non-convertible values)
df1_cleaned = df1_cleaned.dropna(subset=['edition_id'])

gold_winners_df1 = df1_cleaned[df1_cleaned['medal'] == 'Gold']


# Create a pivot table

pivot_gold_winners_df1 = gold_winners_df1.pivot_table(index='country_noc', columns='athlete', values='edition_id')

# gold_winners_df1: This refers to a DataFrame, presumably containing information about gold medal winners 
# in some sports competition. Each row likely represents a gold medalist and their relevant details.

# .pivot_table(...): This is a method provided by pandas that is used to create a pivot table from the given DataFrame. 
# It allows you to reorganize and aggregate the data based on specified columns.

# index='country_noc': This specifies that the 'country_noc' column will be used as the index of the pivot table.
# This means that the unique values in the 'country_noc' column will become the rows of the pivot table.

# columns='athlete': This specifies that the 'athlete' column will be used to create columns in the pivot table. 
# Each unique athlete's name will correspond to a separate column.

# values='edition_id': This specifies that the values to be aggregated in the pivot table will be taken from the 
# 'edition_id' column. This could be an identifier for the edition of the sports competition.

# So, putting it all together, the line of code 
# pivot_gold_winners_df1 = gold_winners_df1.pivot_table(index='country_noc', columns='athlete', values='edition_id') 
# creates a pivot table where each row represents a country, each column represents an athlete, 
# and the values in the table represent the edition IDs of the competition 
# where each athlete from each country won a gold medal.


pivot_gold_winners_df1.head(30)


<ipython-input-76-467ac3dbe162>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1_cleaned['edition_id'] = pd.to_numeric(df1_cleaned['edition_id'], errors='coerce')


athlete       Abdul Rashid   Abelardo   Abner   Adenízia   Alfonso   Alison  \
country_noc                                                                   
ALG                    NaN        NaN     NaN        NaN       NaN      NaN   
ANZ                    NaN        NaN     NaN        NaN       NaN      NaN   
ARG                    NaN        NaN     NaN        NaN       NaN      NaN   
ARM                    NaN        NaN     NaN        NaN       NaN      NaN   
AUS                    NaN        NaN     NaN        NaN       NaN      NaN   
AUT                    NaN        NaN     NaN        NaN       NaN      NaN   
AZE                    NaN        NaN     NaN        NaN       NaN      NaN   
BAH                    NaN        NaN     NaN        NaN       NaN      NaN   
BDI                    NaN        NaN     NaN        NaN       NaN      NaN   
BEL                    NaN        NaN     NaN        NaN       NaN      NaN   
BER                    NaN        NaN     NaN        NaN       NaN      NaN   
BLR                    NaN        NaN     NaN        NaN       NaN      NaN   
BRA                    NaN        NaN    61.0       54.0       NaN     59.0   
BRN                    NaN        NaN     NaN        NaN       NaN      NaN   
BUL                    NaN        NaN     NaN        NaN       NaN      NaN   
CAN                    NaN        NaN     NaN        NaN       NaN      NaN   
CHI                    NaN        NaN     NaN        NaN       NaN      NaN   
CHN                    NaN        NaN     NaN        NaN       NaN      NaN   
CIV                    NaN        NaN     NaN        NaN       NaN      NaN   
CMR                    NaN        NaN     NaN        NaN       NaN      NaN   
COL                    NaN        NaN     NaN        NaN       NaN      NaN   
CRC                    NaN        NaN     NaN        NaN       NaN      NaN   
CRO                    NaN        NaN     NaN        NaN       NaN      NaN   
CUB                    NaN        NaN     NaN        NaN       NaN      NaN   
CZE                    NaN        NaN     NaN        NaN       NaN      NaN   
DEN                    NaN        NaN     NaN        NaN       NaN      NaN   
DOM                    NaN        NaN     NaN        NaN       NaN      NaN   
ECU                    NaN        NaN     NaN        NaN       NaN      NaN   
EGY                    NaN        NaN     NaN        NaN       NaN      NaN   
ESP                    NaN       23.0     NaN        NaN      23.0      NaN   

athlete       Amauri   Anderson   Antony   Brenno  ...  Šarūnas Marčiulionis  \
country_noc                                        ...                         
ALG              NaN        NaN      NaN      NaN  ...                   NaN   
ANZ              NaN        NaN      NaN      NaN  ...                   NaN   
ARG              NaN        NaN      NaN      NaN  ...                   NaN   
ARM              NaN        NaN      NaN      NaN  ...                   NaN   
AUS              NaN        NaN      NaN      NaN  ...                   NaN   
AUT              NaN        NaN      NaN      NaN  ...                   NaN   
AZE              NaN        NaN      NaN      NaN  ...                   NaN   
BAH              NaN        NaN      NaN      NaN  ...                   NaN   
BDI              NaN        NaN      NaN      NaN  ...                   NaN   
BEL              NaN        NaN      NaN      NaN  ...                   NaN   
BER              NaN        NaN      NaN      NaN  ...                   NaN   
BLR              NaN        NaN      NaN      NaN  ...                   NaN   
BRA             23.0       26.0     61.0     61.0  ...                   NaN   
BRN              NaN        NaN      NaN      NaN  ...                   NaN   
BUL              NaN        NaN      NaN      NaN  ...                   NaN   
CAN              NaN        NaN      NaN      NaN  ...                   NaN   
CHI              NaN        NaN

In [77]:
# Copy df to df_n
df_n = df.copy()

# Convert 'weight' column to numeric, replacing non-convertible values with NaN
df_n['weight'] = pd.to_numeric(df_n['weight'], errors='coerce')

# Drop rows where 'weight' is NaN (non-convertible values)
df_n = df_n.dropna(subset=['weight'])

# Convert 'height' column to numeric, replacing non-convertible values with NaN
df_n['height'] = pd.to_numeric(df_n['height'], errors='coerce')

# Drop rows where 'height' is NaN (non-convertible values)
df_n = df_n.dropna(subset=['height'])

# df_n.head()

# Create a pivot table
pivot_df_n = df_n.pivot_table(index='country', columns='height', values='weight')
pivot_df_n.tail()



height                   127.0  128.0  130.0  131.0  132.0  133.0  135.0  \
country                                                                    
 West Indies Federation    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
 Yemen                     NaN    NaN    NaN    NaN    NaN    NaN    NaN   
 Yugoslavia                NaN    NaN    NaN    NaN    NaN    NaN    NaN   
 Zambia                    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
 Zimbabwe                  NaN    NaN    NaN    NaN    NaN    NaN    NaN   

height                   136.0  137.0  138.0  ...  214.0  215.0  216.0  217.0  \
country                                       ...                               
 West Indies Federation    NaN    NaN    NaN  ...    NaN    NaN    NaN    NaN   
 Yemen                     NaN    NaN    NaN  ...    NaN    NaN    NaN    NaN   
 Yugoslavia                NaN    NaN    NaN  ...    NaN    NaN    NaN    NaN   
 Zambia                    NaN    NaN    NaN  ...    NaN    NaN    NaN    NaN   
 Zimbabwe                  NaN    NaN   66.0  ...    NaN    NaN    NaN    NaN   

height                   218.0  219.0  220.0  221.0  223.0  226.0  
country                                                            
 West Indies Federation    NaN    NaN    NaN    NaN    NaN    NaN  
 Yemen                     NaN    NaN    NaN    NaN    NaN    NaN  
 Yugoslavia                NaN    NaN    NaN    NaN    NaN    NaN  
 Zambia                    NaN    NaN    NaN    NaN    NaN    NaN  
 Zimbabwe                  NaN    NaN    NaN    NaN    NaN    NaN  

[5 rows x 95 columns]

###  Group By .groupby() The .groupby() function in pandas is a powerful tool for splitting a DataFrame into groups based on some criteria and then applying a function to each group independently. It's a fundamental operation for data analysis and can be used for various purposes. After creating a grouped object, you can apply various aggregation functions like .sum(), .mean(), .count(), and more to perform calculations on the grouped data.

In [78]:
df.head()

athlete_id                 name     sex        born height weight  \
0       16809       Károly Teppert    Male  1891-07-20     na     na   
1       43737    Andrzej Socharski    Male  1947-08-31  173.0     72   
2       50147  Nathalie Wunderlich  Female  1971-06-03  170.0     50   
3        5085           Miha Lokar    Male  1935-09-10  182.0     76   
4      136329          Austin Hack    Male  1992-05-17  203.0    100   

          country country_noc  \
0         Hungary         HUN   
1          Poland         POL   
2     Switzerland         SUI   
3      Yugoslavia         YUG   
4   United States         USA   

                                         description  \
0  Károly Teppert started competing in cycling in...   
1                                                 na   
2                                                 na   
3                                                 na   
4                                                 na   

                                       special_notes  
0                                                 na  
1  Listed in Olympians Who Won a Medal at the Wor...  
2                                                 na  
3  Listed in Olympians Who Won a Medal at the Eur...  
4                                                 na

In [79]:
import pandas as pd

# Assuming df is your DataFrame

# Convert 'weight' column to numeric, replacing non-convertible values with NaN
df['weight'] = pd.to_numeric(df['weight'], errors='coerce')

# Drop rows where 'weight' is NaN (non-convertible values)
df = df.dropna(subset=['weight'])

# Convert 'height' column to numeric, replacing non-convertible values with NaN
df['height'] = pd.to_numeric(df['height'], errors='coerce')

# Drop rows where 'height' is NaN (non-convertible values)
df = df.dropna(subset=['height'])


# Group by 'country' and calculate the average weight and height
average_stats_by_country = df.groupby('country')[['weight', 'height']].mean()

# groupby('country'): This part of the code groups the data in the DataFrame based on unique 
# values in the 'country' column. This is similar to what we discussed in the previous explanation. 
# It creates separate groups for each unique country.

# [['weight', 'height']]: This part of the code specifies the columns you're interested in, 
# namely 'weight' and 'height'. By using double square brackets ([[]]), you're indicating that you want 
# to select these two columns as a subset of the original DataFrame.

# .mean(): After grouping by country and selecting the 'weight' and 'height' columns, 
# you apply the .mean() method. This calculates the average (mean) value for each column 
# within each group of countries. 
# As a result, you get the average weight and average height for each country separately.

# Display the result
print(average_stats_by_country)

                             weight      height
country                                        
 Afghanistan              67.076923  169.794872
 Albania                  74.285714  174.678571
 Albania  Australia       65.000000  166.000000
 Albania  Bulgaria       100.000000  189.000000
 Albania  Greece          64.000000  165.000000
...                             ...         ...
 West Indies Federation   71.500000  176.000000
 Yemen                    60.476190  168.952381
 Yugoslavia               76.662484  180.233375
 Zambia                   64.969072  172.092784
 Zimbabwe                 67.873626  172.983516

[645 rows x 2 columns]


In [80]:
# Calculate multiple aggregation functions on specific columns
aggregated_stats = df.agg({
    'weight': ['mean', 'median', 'max'],
    'height': ['mean', 'median', 'max']
})

# Display the aggregated statistics
aggregated_stats.head()

# .agg(aggregations):The agg method is used to perform these aggregations. 
# The aggregations variable should be a dictionary that defines which aggregation functions 
# to apply to specific columns in the grouped data. An aggregation function can be anything that 
# summarizes a group of values, such as mean, sum, count, etc.

weight      height
mean     71.869244  176.316269
median   70.000000  176.000000
max     198.000000  226.000000

In [81]:
# Group by 'country' and calculate multiple aggregation functions on specific columns
aggregations = {
    'weight': ['mean', 'median', 'max'],
    'height': ['mean', 'median', 'max']
}

grouped_and_aggregated = df.groupby('country').agg(aggregations)

# groupby('country'): This is a pandas method that is used to group the data in the DataFrame 
# based on unique values in the 'country' column. When you apply a groupby operation, you're essentially 
# creating groups of data where each group contains rows with the same value in the 'country' column.

# .agg(aggregations): After the data is grouped by the 'country' column, you're applying aggregation 
# functions to each group. The agg method is used to perform these aggregations. 
# The aggregations variable should be a dictionary that defines which aggregation functions 
# to apply to specific columns in the grouped data. An aggregation function can be anything that 
# summarizes a group of values, such as mean, sum, count, etc.

# Display the aggregated statistics
grouped_and_aggregated.head(30)


weight                    height  \
                                            mean median    max        mean   
country                                                                      
 Afghanistan                           67.076923   66.0  111.0  169.794872   
 Albania                               74.285714   74.5  130.0  174.678571   
 Albania  Australia                    65.000000   65.0   65.0  166.000000   
 Albania  Bulgaria                    100.000000  100.0  100.0  189.000000   
 Albania  Greece                       64.000000   64.0   64.0  165.000000   
 Albania  Kosovo                       52.000000   52.0   52.0  162.000000   
 Albania  United States                57.000000   57.0   57.0  163.000000   
 Algeria                               68.817901   67.0  120.0  175.311728   
 Algeria  France                       66.000000   66.0   67.0  172.000000   
 American Samoa                        84.894737   82.0  127.0  176.578947   
 Andorra                               69.170213   66.0   90.0  173.659574   
 Angola                                74.480000   72.0  114.0  178.480000   
 Antigua and Barbuda                   67.410714   68.0   91.0  175.821429   
 Argentina                             73.998401   73.0  140.0  176.993605   
 Argentina  Australia                  91.000000   91.0   91.0  199.000000   
 Argentina  Italy                      70.000000   70.0   70.0  175.500000   
 Argentina  Netherlands                92.000000   92.0   92.0  180.000000   
 Argentina  Spain                      52.000000   52.0   52.0  170.000000   
 Argentina  Ukraine                    60.000000   60.0   60.0  167.000000   
 Armenia                               72.219512   69.0  150.0  172.276423   
 Armenia  Belarus                      84.000000   84.0   84.0  186.000000   
 Armenia  Bulgaria                     60.000000   60.0   60.0  162.000000   
 Armenia  Kazakhstan                   63.000000   63.0   63.0  166.000000   
 Armenia  Russian Federation           91.000000   91.0  100.0  184.000000   
 Armenia  Soviet Union                 66.000000   69.0   72.0  172.333333   
 Armenia  Soviet Union  Unified Team   74.000000   74.0   78.0  169.500000   
 Aruba                                 72.750000   68.5  150.0  174.642857   
 Aruba  Individual Olympic Athletes    65.000000   65.0   65.0  171.000000   
 Aruba  Netherlands Antilles           63.000000   63.0   65.0  168.500000   
 Australasia                           75.750000   75.5   85.0  179.500000   

                                                    
                                     median    max  
country                                             
 Afghanistan                          170.0  186.0  
 Albania                              175.0  195.0  
 Albania  Australia                   166.0  166.0  
 Albania  Bulgaria                    189.0  189.0  
 Albania  Greece                      165.0  165.0  
 Albania  Kosovo                      162.0  162.0  
 Albania  United States               163.0  163.0  
 Algeria                              175.5  200.0  
 Algeria  France                      172.0  175.0  
 American Samoa                       175.0  188.0  
 Andorra                              174.0  189.0  
 Angola                               177.0  208.0  
 Antigua and Barbuda                  175.5  192.0  
 Argentina                            176.0  210.0  
 Argentina  Australia                 199.0  199.0  
 Argentina  Italy                     175.5  181.0  
 Argentina  Netherlands               180.0  180.0  
 Argentina  Spain                     170.0  170.0  
 Argentina  Ukraine                   167.0  167.0  
 Armenia                              171.0  195.0  
 Armenia  Belarus                     186.0  186.0  
 Armenia  Bulgaria                    162.0  162.0  
 Armenia  Kazakhstan                  166.0  166.0  
 Armenia  Russian Federation          184.0  190.0  
 Armenia  Soviet Union              